core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -403,42 +406,79 @@ private void logFailedEvents(final List<BatchItem> batch) {
         LOG.info("Failed events in batch {}", sb.toString());
     }
 
-    public void syncPostEvent(final NakadiRecord nakadiRecord) {
+    /**
+     * The method sends list of events to Kafka and waiting for the result of each event.
+     *
+     * @param nakadiRecords list of the events to publish
+     * @return empty list if no errors otherwise list with the errored events
+     */
+    public List<NakadiRecordMetadata> sendEvents(final String topic,
+                                                 final List<NakadiRecord> nakadiRecords) {
         final Producer<byte[], byte[]> producer = kafkaFactory.takeProducer();
+        final CountDownLatch latch = new CountDownLatch(nakadiRecords.size());
+        final List<NakadiRecordMetadata> errors = new LinkedList<>();
+        int publishingEventIdx = 0;
         try {
-            final ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(
-                    nakadiRecord.getTopic(),
-                    nakadiRecord.getPartition(),
-                    nakadiRecord.getEventKey(),
-                    nakadiRecord.getData());
-
-            // fixme avro do not forget EventOwnerHeader, set via metadata?
-            // for the internal KPI evnets it is fine not to have it
-
-            producerRecord.headers().add(
-                    NakadiRecord.HEADER_FORMAT,
-                    nakadiRecord.getFormat());
-
-            producer.send(producerRecord, ((metadata, exception) -> {
-                if (null != exception) {
-                    LOG.warn("Failed to publish to kafka topic '{}' event-type '{}'",
-                            nakadiRecord.getTopic(), nakadiRecord.getEventType(), exception);
-                    if (isExceptionShouldLeadToReset(exception)) {
-                        kafkaFactory.terminateProducer(producer);
+            for (final NakadiRecord nakadiRecord : nakadiRecords) {
+                publishingEventIdx++;
+                final ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(
+                        topic,
+                        nakadiRecord.getPartition(),
+                        nakadiRecord.getEventKey(),
+                        nakadiRecord.getData());
+
+                // fixme avro do not forget EventOwnerHeader, set via metadata?
+                // for the internal KPI evnets it is fine not to have it
+
+                producerRecord.headers().add(
+                        NakadiRecord.HEADER_FORMAT,
+                        nakadiRecord.getFormat());
+                producer.send(producerRecord, ((metadata, exception) -> {
+                    latch.countDown();
+                    if (null != exception) {
+                        errors.add(new NakadiRecordMetadata(
+                                nakadiRecord.getEventType(),
+                                nakadiRecord.getPartition(),
+                                nakadiRecord.getData(),
```

Comment:
The parameter name is `metadata`.  Not sure I understand what's happening here.

Reply:
thanks, I will check that